File: Python/1062-talent_and_virtue.py


class People():

    high_line = 0

    def __init__(self, id_num, virtue, talent):
        self.id_num = id_num
        self.virtue = virtue
        self.talent = talent
        self.total_value = virtue + talent

        if self.virtue >= People.high_line:
            if self.talent >= People.high_line:
                self.type = 1  # sage
            else:
                self.type = 2  # nobleman
        else:
            if self.virtue >= self.talent:
                self.type = 3  # fool man
            else:
                self.type = 4  # small man

    def __lt__(self, other):
        if self.type < other.type:
            return True
        elif self.type > other.type:
            return False
        # rank by total_value
        if self.total_value > other.total_value:
            return True
        elif self.total_value < other.total_value:
            return False
        # rank by virtue
        if self.virtue > other.virtue:
            return True
        elif self.virtue < other.virtue:
            return False
        # rank by ID
        return self.id_num < other.id_num

    def __str__(self):
        return f'{self.id_num} {self.virtue} {self.talent}'


if __name__ == '__main__':
    n_people, low_line, high_line = [int(i) for i in input().split()]
    # set high_line
    People.high_line = high_line
    all_people = []
    for _ in range(n_people):
        id_num, virtue, talent = input().split()
        virtue = int(virtue)
        talent = int(talent)
        if virtue < low_line or talent < low_line:
            continue
        all_people.append(People(id_num, virtue, talent))
    # sorting
    all_people.sort()
    print(len(all_people))
    for people in all_people:
        print(people)

File: C++/1062-talent_and_virtue.cc
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>


int high_line = 80;  // default to 80


struct People {
    // constructors
    People() = default;
    People(std::string id_num, int virtue, int talent):
        id_num(id_num), virtue(virtue), talent(talent), total_value(virtue + talent)
    {
        if (virtue >= high_line) {
            if (talent >= high_line)
                type = 1;  // sage
            else
                type = 2;  // nobleman
        } else {
            if (virtue >= talent)
                type = 3;  // fool man
            else
                type = 4;  // small man
        }   
    }
    // members
    std::string id_num;
    int virtue, talent, total_value, type;
};


bool operator<(const People &p1, const People &p2) {
    if (p1.type < p2.type)
        return true;
    else if (p1.type > p2.type)
        return false;
    // rank by total_value
    if (p1.total_value > p2.total_value)
        return true;
    else if (p1.total_value < p2.total_value)
        return false;
    // rank by virtue
    if (p1.virtue > p2.virtue)
        return true;
    else if (p1.virtue < p2.virtue)
        return false;
    // rank by id_num
    return p1.id_num < p2.id_num;
}


std::ostream &operator<<(std::ostream &os, const People &p) {
    os << p.id_num << ' ' << p.virtue << ' ' << p.talent;
    return os;
}


int main() {

    std::ios::sync_with_stdio(false);

    int n_people, low_line;
    std::cin >> n_people >> low_line;
    // set high_line
    std::cin >> high_line;

    std::vector<People> all_people;
    while (n_people--) {
        std::string id_num;
        int virtue, talent;
        std::cin >> id_num >> virtue >> talent;
        if (virtue < low_line || talent < low_line) continue;
        all_people.push_back(People(id_num, virtue, talent));
    }

    std::sort(all_people.begin(), all_people.end());
    std::cout << all_people.size() << '\n';
    for (auto &people : all_people) {
        std::cout << people << '\n';
    }

    return 0;
}
